ext/har/types.go

```diff
@@ -0,0 +1,365 @@
+}
+
+type Creator struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Comment string `json:"comment,omitempty"`
+}
+
+type Browser struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Comment string `json:"comment,omitempty"`
+}
+
+type Page struct {
+	ID              string      `json:"id,omitempty"`
+	StartedDateTime time.Time   `json:"startedDateTime"`
+	Title           string      `json:"title"`
+	PageTimings     PageTimings `json:"pageTimings"`
+	Comment         string      `json:"comment,omitempty"`
+}
+
+type Entry struct {
+	PageRef         string    `json:"pageref,omitempty"`
+	StartedDateTime time.Time `json:"startedDateTime"`
+	Time            int64     `json:"time"`
+	Request         *Request  `json:"request"`
+	Response        *Response `json:"response"`
+	Cache           Cache     `json:"cache"`
+	Timings         Timings   `json:"timings"`
+	ServerIpAddress string    `json:"serverIpAddress,omitempty"`
+	Connection      string    `json:"connection,omitempty"`
+	Comment         string    `json:"comment,omitempty"`
+}
+
+type Cache struct {
+	BeforeRequest *CacheEntry `json:"beforeRequest,omitempty"`
+	AfterRequest  *CacheEntry `json:"afterRequest,omitempty"`
+}
+
+type CacheEntry struct {
+	Expires    string `json:"expires,omitempty"`
+	LastAccess string `json:"lastAccess"`
+	ETag       string `json:"eTag"`
+	HitCount   int    `json:"hitCount"`
+	Comment    string `json:"comment,omitempty"`
+}
+
+type Request struct {
+	Method      string          `json:"method"`
+	Url         string          `json:"url"`
+	HttpVersion string          `json:"httpVersion"`
+	Cookies     []Cookie        `json:"cookies"`
+	Headers     []NameValuePair `json:"headers"`
+	QueryString []NameValuePair `json:"queryString"`
+	PostData    *PostData       `json:"postData,omitempty"`
+	BodySize    int64           `json:"bodySize"`
+	HeadersSize int64           `json:"headersSize"`
+}
+
+func ParseRequest(req *http.Request, captureContent bool) *Request {
+	if req == nil {
+		return nil
+	}
+	harRequest := Request{
+		Method:      req.Method,
+		Url:         req.URL.String(),
+		HttpVersion: req.Proto,
+		Cookies:     parseCookies(req.Cookies()),
+		Headers:     parseStringArrMap(req.Header),
+		QueryString: parseStringArrMap((req.URL.Query())),
+		BodySize:    req.ContentLength,
+		HeadersSize: calcHeaderSize(req.Header),
+	}
+
+	if captureContent && (req.Method == "POST" || req.Method == "PUT") {
+		harRequest.PostData = parsePostData(req)
+	}
+
+	return &harRequest
+}
+
+func (harEntry *Entry) FillIPAddress(req *http.Request) {
+	host, _, err := net.SplitHostPort(req.URL.Host)
+	if err != nil {
+		host = req.URL.Host
+	}
+	if ip := net.ParseIP(host); ip != nil {
+		harEntry.ServerIpAddress = string(ip)
+	}
+
+	if ipaddr, err := net.LookupIP(host); err == nil {
+		for _, ip := range ipaddr {
+			if ip.To4() != nil {
+				harEntry.ServerIpAddress = ip.String()
+				return
+			}
+		}
+	}
```

Comment:
Since this is an optional field, avoid to manually make a DNS resolution and handle only the previous case where we directly have the IP address.
If we manually resolve the IP, the result can be different from the originally resolved IP and will cause confusion.